jose/x25519_test.go

```
@@ -0,0 +1,360 @@
+			Audience:  []string{"test-aud"},
+			Expiry:    NewNumericDate(time.Unix(1234, 0)),
+			NotBefore: NewNumericDate(time.Unix(1200, 0)),
+			IssuedAt:  NewNumericDate(time.Unix(1000, 0)),
+			ID:        "test-jti",
+		}, false},
+		{"ok ed25519", args{
+			SigningKey{Algorithm: EdDSA, Key: edPriv},
+			new(SignerOptions).WithType("JWT"),
+			[]byte(`{"iss":"test-iss","sub":"test-sub","aud":"test-aud","exp":1234,"nbf":1200,"iat":1000,"jti":"test-jti"}`),
+			edPub,
+		}, Claims{
+			Issuer:    "test-iss",
+			Subject:   "test-sub",
+			Audience:  []string{"test-aud"},
+			Expiry:    NewNumericDate(time.Unix(1234, 0)),
+			NotBefore: NewNumericDate(time.Unix(1200, 0)),
+			IssuedAt:  NewNumericDate(time.Unix(1000, 0)),
+			ID:        "test-jti",
+		}, false},
+		{"fail algorithm", args{
+			SigningKey{Algorithm: EdDSA, Key: priv},
+			new(SignerOptions).WithType("JWT"),
+			[]byte(`{"iss":"test-iss","sub":"test-sub","aud":"test-aud","exp":1234,"nbf":1200,"iat":1000,"jti":"test-jti"}`),
+			pub,
+		}, Claims{}, true},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			signer, err := NewSigner(tt.args.sig, tt.args.opts)
+			if (err != nil) != tt.wantErr {
+				t.Errorf("NewSigner() error = %v, wantErr %v", err, tt.wantErr)
+				return
+			}
+			if !tt.wantErr {
+				var got Claims
+
+				sig, err := signer.Sign(tt.args.payload)
+				if err != nil {
+					t.Errorf("Signer.Sign() error = %v", err)
+					return
+				}
+				jwt, err := ParseSigned(sig.FullSerialize())
+				if err != nil {
+					t.Errorf("ParseSigned() error = %v", err)
+					return
+				}
+				if err := Verify(jwt, tt.args.publicKey, &got); err != nil {
+					t.Errorf("Verify() error = %v", err)
+					return
+				}
+				if !reflect.DeepEqual(got, tt.want) {
+					t.Errorf("JSONWebSignature.CompactSerialize() = %v, want %v", got, tt.want)
+				}
+			}
+		})
+	}
+}
+
+func TestX25519Signer_Public(t *testing.T) {
+	pub, priv, err := x25519.GenerateKey(rand.Reader)
```

While investigating the deterministic output for `rand.Reader`, I saw that the Go stdlib implementation of `ed25519.GenerateKey` checks if `rand.Reader` is nil, and if so, will fallback to `crypto/rand.Reader`. It might be nice to mirror that approach for `x25519.GenerateKey`, unless you expect problems down the road?

---

I prefer not to do that, as it can hide other errors, it will probably panic if it's nil.